examples/ppo/main.py

```diff
@@ -0,0 +1,295 @@
+    values = values[:, 0] # convert shapes: (batch, 1) to (batch, )
+    probs = jnp.exp(log_probs)
+    entropy = jnp.sum(-probs*log_probs, axis=1).mean()
+    log_probs_act_taken = jax.vmap(lambda lp, a: lp[a])(log_probs, actions)
+    ratios = jnp.exp(log_probs_act_taken - old_log_probs)
+    # adv. normalization (following the OpenAI baselines)
+    advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+    PG_loss = ratios * advantages
+    clipped_loss = advantages * jax.lax.clamp(1. - clip_param, ratios,
+                                               1. + clip_param)
+    assert(PG_loss.shape == clipped_loss.shape)
+    PPO_loss = -jnp.mean(jnp.minimum(PG_loss, clipped_loss), axis=0)
+    assert(values.shape == returns.shape)
+    value_loss = jnp.mean(jnp.square(returns - values), axis=0)
+    return PPO_loss + vf_coeff*value_loss - entropy_coeff*entropy
+
+  batch_size = BATCH_SIZE
+  iterations = trn_data[0].shape[0] // batch_size
+  trn_data = jax.tree_map(
+    lambda x: x.reshape((iterations, batch_size) + x.shape[1:]), trn_data)
+  loss = 0.
+  for batch in zip(*trn_data):
+    grad_fn = jax.value_and_grad(loss_fn)
+    l, grad = grad_fn(optimizer.target, batch, clip_param, vf_coeff,
+    entropy_coeff)
+    loss += l
+    optimizer = optimizer.apply_gradient(grad)
+    grad_norm = sum(jnp.square(g).sum() for g in jax.tree_leaves(grad))
+  return optimizer, loss, grad_norm
+
+
+def thread_inference(
+  q1 : Queue,
+  q2: Queue,
+  simulators : List[RemoteSimulator],
+  steps_per_actor : int):
+  """Worker function for a separate thread used for inference and running
+  the simulators in order to maximize the GPU/TPU usage. Runs
+  `steps_per_actor` time steps of the game for each of the `simulators`.
+  """
+
+  while(True):
+    optimizer, step = q1.get()
+    all_experience = []
+    for _ in range(steps_per_actor + 1): # +1 to get one more value
+                                         # needed for GAE
+      states = []
+      for sim in simulators:
+        state = sim.conn.recv()
+        states.append(state)
+      states = onp.concatenate(states, axis=0)
+
+      # perform inference
+      # policy_optimizer, step = q1.get()
+      log_probs, values = policy_action(optimizer.target, states)
+
+      probs = onp.exp(onp.array(log_probs))
+      # print("probs after onp conversion", probs)
+
+      for i, sim in enumerate(simulators):
+        # probs[i] should sum up to 1, but there are float round errors
+        # if using jnp.array directly, it required division by probs[i].sum()
+        # better solutions can be thought of
+        # issue might be a result of the network using jnp.int 32, , not 64
+        probabilities = probs[i] # / probs[i].sum()
+        action = onp.random.choice(probs.shape[1], p=probabilities)
+        #in principle, one could avoid sending value and log prob back and forth
+        sim.conn.send((action, values[i, 0], log_probs[i][action]))
```

Comment:
Should it just send the action?

Reply:
Implemented as https://github.com/google/flax/pull/470/commits/d7a8fa45f98200bdc7ac113c5020d3d28fef49f7, thanks for the suggestion!